doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+1. `import A(N=1).*`  adds a `myModule::A` namespace
+2. `import A(N=1) as A1` adds a `myModule::A1` namespace
+
+##### Namespaces for exports
+
+Exports have a particularity: they can remove namespaces. Consider the example:
+```bluespec
+module A {
+  const N: int
+  val a = N
+}
+
+module B {
+  import A(N=1) as A1
+  export A1.*
+}
+```
+
+In this example, `export A1.*` is taking the definitions from `A1` and making them available without any additional namespaces.
+
+This scenario is tricky and, as discussed in a Quint meeting, we could not support it in the first version. Current flattening does support this, but it is not able to distinguish different instances, so problems arise. We are not sure how useful and clear this scenario is for users.
+
+So, exports that change the namespaces of the previously imported definitions are not supported in the first version of the new flattener. We should raise a proper error when an export like that is found. The example above could be rewritten as:
+```bluespec
+module A {
+  const N: int
+  val a = N
+}
+
+module B {
+  import A(N=1) as A1
+  export A1
+}
+```
+
+Currently, this prevents us from using `A1` without a namespace, since we can't write an export for `import A(N=1).*`. @konnov suggested the introduction of `export *` for this end, which is still in consideration.
+
+#### `importedFrom`
+
+This is a simple reference to the import/export/instance statement that created a definition, also gathered during name collection. Used in flattening to distinguish definitions that come from instances from definitions that come from imports/exports.
+
+### Instance Flattener
+
+The instance flattener gets rid of instances, which is the most complicated part of flattening. We cannot use the same id for the same definition of different instances, as they might assume different values[^1].
+
+Each instance statement is replaced with an import statement and generates a new module. The new module has the constants replaced with `pure val` definitions (from the overrides in the instance statement), and all its names are fully-qualified, constructed with the previously defined namespace algorithm. The new import statement, replacing the instance statement, is a unqualified import of the new module.
+
+The names in the module containing the instance statement also have to be updated with a namespace, since they might refer to definitions that now include a namespace. We look at each name in the module and, if it refers to a definition that comes from an instance, (that is, its `importFrom` field has an instance), then we replace the name with a namespaced version of it (using the `namespaces` field). For this to work properly, any existing `import` and `export` statements should have been flattened before the instance flattener is run, ensuring that the definitions referred to by these names already have their proper namespaces.
```

Comment:
Probably connected to my question above – I don't follow why this is necessary.
Could you give an example to illustrate?

Reply:
Yes! So, this happens "after"[^1] we generate a new module for each instance and replace the instances with imports. 

Consider the original modules (is the same from the example section but without `C` to simplify)

```bluespec
module A {
  const N: int 
  val a = N
}

module B {
  import A(N=1) as A1
  val b = A1::a
}
```

Imagine we've created the new module (`B::A1`) for the instance statement `import A(N=1) as A1` in `B`, and replaced that statement with `import B::A1.*`:

```bluespec
module A {
  const N: int
  val a = N
}

module B::A1 {
  pure val B::A1::N = 1
  pure val B::A1::a = B::A1::N
}

module B {
  import B::A1.*
  val b = A1::a
}
```

Now we have a problem in the last definition here: `val b = A1::a`, because `A1::a` doesn't exist anymore, and the name for that is `B::A1::a`, so we have to fix it. This is what the paragraph is describing.

One approach that might seem like a good idea looking at this is what I thought about in the quint meeting last week (and also when I was first implementing this). Don't add the namespaces to the definitions in the new module, and import them with a qualifier: 

```bluespec
module A {
  const N: int
  val a = N
}

module B::A1 {
  pure val N = 1
  pure val a = N
}

module B {
  import B::A1 as A1
  val b = A1::a
}
```

That looks simpler here, but it results in name collisions in some of our examples. 

[^1]: It is actually at the same time, but imagine it's after since this is a consequence of the other thing

Reply:
> That looks simpler here, but it results in name collisions in some of our examples.

Ah – I guess that's the same thing that confused me here: https://github.com/informalsystems/quint/pull/1087#discussion_r1285617259

But I trust you if you say it's not possible this way.